src/Compilers/Core/Portable/SourceGeneration/Nodes/InputNode.cs

```diff
@@ -26,5 +28,40 @@ public NodeStateTable<T> UpdateStateTable(DriverStateTable.Builder graphState, N
         // PROTOTYPE(source-generators): how does this work? we definitly want to be able to add custom comparers to the input nodes
         // I guess its just a 'compare only' node with this as the input?
         public IIncrementalGeneratorNode<T> WithComparer(IEqualityComparer<T> comparer) => this;
+
+        public NodeStateTable<T> GetInputTable(NodeStateTable<T> previousTable, T item)
+        {
+            // PROTOTYPE(source-generators): we should compare the values, not just assume they were added
+            return NodeStateTable<T>.WithSingleItem(item, EntryState.Added);
+        }
+
+        public NodeStateTable<T> GetInputTable(NodeStateTable<T> previousTable, IEnumerable<T> items)
+        {
+            // create a mutable hashset of the new items we can check against
+            PooledHashSet<T> itemsSet = PooledHashSet<T>.GetInstance();
+            foreach (var item in items)
+            {
+                itemsSet.Add(item);
+            }
+
+            var builder = new NodeStateTable<T>.Builder();
+
+            // for each item in the previous table, check if its still in the new items
+            foreach ((var oldItem, _) in previousTable)
+            {
+                bool inItemSet = itemsSet.Remove(oldItem);
+                builder.AddEntriesFromPreviousTable(previousTable, inItemSet ? EntryState.Cached : EntryState.Removed);
+            }
+
+            // any remaining new items are added
+            foreach (var newItem in itemsSet)
```

Comment:
>newItem in itemsSet [](http://example.com/codeflow?start=25&length=19)

The new items are not necessarily added in the order they were passed in `items`. Is that an issue?

Reply:
No, it wont matter, because this is the 'root' node of the chain. Everything downstream will get the same order we set here, and they're new nodes so nothing is depending on them yet.

---
In reply to: [627713200](https://github.com/dotnet/roslyn/pull/53147#discussion_r627713200) [](ancestors = 627713200)